### PR TITLE
Rocksdb checkpoint

### PR DESCRIPTION
**Pull request by aaudiber**

### CODE REVIEW

**Comment by aaudiber** [comment]
Jenkins, test this please

**Comment by aaudiber** [comment]
Jenkins, test this please

File: core/server/master/src/main/java/alluxio/master/file/activesync/ActiveSyncManager.java [hunk 1]
```diff
@@ -708,4 +692,45 @@ public void recoverFromStartSync(AlluxioURI uri, long mountId) {
...
+      apply(entry.getAddSyncPoint());
+      return true;
+    } else if (entry.hasRemoveSyncPoint()) {
+      apply(entry.getRemoveSyncPoint());
+      return true;
+    } else if (entry.hasActiveSyncTxId()) {
+      File.ActiveSyncTxIdEntry activeSyncTxId = entry.getActiveSyncTxId();
+      setTxId(activeSyncTxId.getMountId(), activeSyncTxId.getTxId());
+      return true;
+    }
```
Comment (calvinjia) [inline core/server/master/src/main/java/alluxio/master/file/activesync/ActiveSyncManager.java]: Any good way to guarantee we have all the cases?

File: core/server/master/src/main/java/alluxio/master/file/activesync/ActiveSyncManager.java [hunk 2]
```diff
@@ -708,4 +692,45 @@ public void recoverFromStartSync(AlluxioURI uri, long mountId) {
...
+  public void resetState() {
+    for (long mountId : mFilterMap.keySet()) {
+      try {
+        // stops sync point under this mount point. Note this clears the sync point and
+        // stops associated polling threads.
+        stopSyncForMount(mountId);
+      } catch (IOException e) {
+        LOG.info("reset: IOException resetting mountId {}, exception {}", mountId, e);
+      } catch (InvalidPathException e) {
+        LOG.info("reset: InvalidPathException resetting mountId {}, exception {}", mountId, e);
```
Comment (calvinjia) [inline core/server/master/src/main/java/alluxio/master/file/activesync/ActiveSyncManager.java]: Why prefix with `reset`?

File: core/server/master/src/main/java/alluxio/master/file/meta/InodeTree.java [hunk 3]
```diff
@@ -219,6 +215,11 @@ public InodeTree(InodeStore inodeStore, ContainerIdGenerable containerIdGenerato
     mMountTable = mountTable;
   }
 
+  @Override
+  public String getName() {
+    return mState.getName();
```
Comment (calvinjia) [inline core/server/master/src/main/java/alluxio/master/file/meta/InodeTree.java]: Does this cause a problem later if we need to add something other than `mState` (as unlikely as that is)

File: core/server/master/src/main/java/alluxio/master/file/meta/InodeTreePersistentState.java [hunk 4]
```diff
@@ -641,12 +613,91 @@ private long getIdFromPath(Path path) {
...
+    } else if (entry.hasAsyncPersistRequest()) {
+      applyAsyncPersist(entry.getAsyncPersistRequest());
+    } else if (entry.hasCompleteFile()) {
+      applyCompleteFile(entry.getCompleteFile());
+    } else if (entry.hasInodeLastModificationTime()) {
+      applyInodeLastModificationTime(entry.getInodeLastModificationTime());
+    } else if (entry.hasPersistDirectory()) {
+      applyPersistDirectory(entry.getPersistDirectory());
+    } else if (entry.hasSetAttribute()) {
+      applySetAttribute(entry.getSetAttribute());
```
Comment (calvinjia) [inline core/server/master/src/main/java/alluxio/master/file/meta/InodeTreePersistentState.java]: Do we need to keep the deprecated ones for backward compatibility?

File: core/server/master/src/main/java/alluxio/master/file/meta/InodeTreePersistentState.java [hunk 5]
```diff
@@ -641,12 +613,91 @@ private long getIdFromPath(Path path) {
...
+  @Override
+  public void resetState() {
     mInodeStore.clear();
     mReplicationLimitedFileIds.clear();
     mPinnedInodeFileIds.clear();
   }
+
+  @Override
+  public void writeToCheckpoint(OutputStream output) throws IOException, InterruptedException {
+    mInodeStore.writeToCheckpoint(output);
```
Comment (calvinjia) [inline core/server/master/src/main/java/alluxio/master/file/meta/InodeTreePersistentState.java]: Do we need to write the other stateful objects?

File: core/server/common/src/main/java/alluxio/master/journal/JournalReader.java [hunk 6]
```diff
@@ -39,4 +49,31 @@
    * @return the next sequence number
    */
   long getNextSequenceNumber();
+
+  /**
+   * Advances the reader to the next element, and returns what's next.
```
[the remark this answers is not in the document]
Comment (aaudiber) [reply core/server/common/src/main/java/alluxio/master/journal/JournalReader.java]: done

File: core/server/common/src/main/java/alluxio/master/journal/JournalReader.java [hunk 7]
```diff
@@ -28,9 +28,19 @@
...
+  /**
+   * Returns an input stream for reading a checkpoint.
+   *
+   * Users should always call getCheckpoint() before calling read() to make sure there isn't a
+   * checkpoint available.
+   *
+   * @return the input stream for reading the checkpoint, or empty if there is no available
+   *         checkpoint
+   */
+  InputStream getCheckpoint();
```
[the remark this answers is not in the document]
Comment (aaudiber) [reply core/server/common/src/main/java/alluxio/master/journal/JournalReader.java]: updated to make `getCheckpoint`/`getEntry` idempotent

File: core/server/common/src/main/java/alluxio/master/journal/ufs/UfsJournalReader.java [hunk 8]
```diff
@@ -247,4 +258,23 @@ private void updateInputStream() throws IOException {
...
+  @Override
+  public State advance() throws IOException {
+    mCheckpointStream = null;
+    mNextEntry = null;
+    while (true) {
+      updateInputStream();
+      if (mCheckpointStream != null) {
+        return State.CHECKPOINT;
+      }
+      advanceEntry();
```
[the remark this answers is not in the document]
Comment (aaudiber) [reply core/server/common/src/main/java/alluxio/master/journal/ufs/UfsJournalReader.java]: `update*` is for methods which don't move the reader forward, while `advance*` is for methods that do

File: core/server/master/src/main/java/alluxio/master/file/activesync/ActiveSyncManager.java (hunk 1 above)
Comment (aaudiber) [reply core/server/master/src/main/java/alluxio/master/file/activesync/ActiveSyncManager.java]: same as before, unrecognized entries will cause a runtime exception to be thrown

File: core/server/master/src/main/java/alluxio/master/file/activesync/ActiveSyncManager.java (hunk 2 above)
Comment (aaudiber) [reply core/server/master/src/main/java/alluxio/master/file/activesync/ActiveSyncManager.java]: removed the prefix (this code was actually just moved)

**Pull request closed without merging**